# (6주차) 2월18일 (1)
> 16-다차원 리스트 
- toc: true
- branch: master
- badges: false
- comments: true
- author: dinonene
- categories: [python]

`-` 소개

`-` 2차원 리스트 처리

`-` 2차원 리스트와 함수

`-` 예제

`-` 다차원 리스트

### 소개

테이블이나 행렬은 2차원 리스트로 표현할 수 있다.



`-` 대한민국 도시들 간 거리



|   |서울|부산|대구|광주|
|---|----|---|----|----|
|서울|0  |325|237|267|
|부산|325|0|87|202|
|대구|237|87|0|172|
|광주|267|202|172|0|

In [1]:
distance = [[0, 325, 237, 267],
            [325, 0, 87, 202],
            [237, 87, 0, 172],
            [267, 202, 172, 0]]

### 2차원 리스트 처리하기

####  2차원 리스트 표현

In [9]:
matrix_ = [[1,2,3,4,5],
           [6,7,8,9,10],
           [11,12,13,14,15]]

`-` 각 요소는 두 개의 첨자를 이용하여 표현

In [10]:
matrix_[0][2]

3

In [11]:
matrix_[2][3]

14

In [12]:
matrix_[1][1]

7

#### 2차원 리스트 초기화

`-` 사용자 입력 값으로 초기화

`-` 무작위 값으로 초기화

In [33]:
matrix = []
for row in range(3): ## number of rows = 3
    matrix.append([])
    for col in range(2): ## number of columns = 2
        value = eval(input("value:"))
        matrix[row].append(value)

value: 1
value: 2
value: 3
value: 4
value: 5
value: 6


In [36]:
matrix ## 3행 2열의 2차원 리스트

[[1, 2], [3, 4], [5, 6]]

In [46]:
import random
matrix = []
numberOfRows=3
numberOfColumns=2
for row in range(numberOfRows):
    matrix.append([])
    #print(matrix)
    for col in range(numberOfColumns):
        matrix[row].append(random.randint(0,99))
        #print(matrix)

[[]]
[[51]]
[[51, 15]]
[[51, 15], []]
[[51, 15], [47]]
[[51, 15], [47, 46]]
[[51, 15], [47, 46], []]
[[51, 15], [47, 46], [79]]
[[51, 15], [47, 46], [79, 98]]


In [47]:
matrix

[[51, 15], [47, 46], [79, 98]]

#### 2차원 리스트 출력하기, 섞기, 정렬

`-` 2차원 리스트 출력

In [66]:
len(matrix), len(matrix[0])

(3, 2)

In [68]:
for row in range(len(matrix)):
    for col in range(len(matrix[row])):
        print(matrix[row][col], end=' ')
    print()

51 15 
47 46 
79 98 


`-` 2차원 리스트 섞기

In [70]:
matrix

[[51, 15], [47, 46], [79, 98]]

In [76]:
for row in range(len(matrix)):
    for col in range(len(matrix[row])):
        i = random.randint(0, len(matrix)-1)
        j = random.randint(0, len(matrix[row])-1)
        matrix[row][col], matrix[i][j] = matrix[i][j], matrix[row][col]

In [77]:
matrix

[[79, 98], [46, 15], [47, 51]]

- 리스트 안의 원소들이 잘 섞여진 것을 확인할 수 있다.

#### 2차원 리스트 처리

`-` 모든 원소들의 합 구하기

In [78]:
matrix

[[79, 98], [46, 15], [47, 51]]

In [79]:
total = 0
for row in matrix:
    for value in row:
        total += value
        print(total)

79
177
223
238
285
336


In [80]:
total

336

`-` 각 열의 합 구하기

In [82]:
matrix

[[79, 98], [46, 15], [47, 51]]

In [81]:
for col in range(len(matrix[0])):
    total = 0
    for row in range(len(matrix)):
        total += matrix[row][col]
    print("Sum of column", col, "is", total)

Sum of column 0 is 172
Sum of column 1 is 164


In [83]:
79+46+47 # 1열 합

172

In [84]:
98+15+51 # 2열 합

164

- 각 열의 합이 잘 계산되었다.

`-` 합이 가장 큰 행 구하기

In [85]:
matrix

[[79, 98], [46, 15], [47, 51]]

In [92]:
maxRow = sum(matrix[0])
indexRow = 0
for row in range(1, len(matrix)):
    if sum(matrix[row]) > maxRow:
        maxRow = sum(matrix[row])
        indexRow = row
print("Row", indexRow, "has max sum of " , maxRow)

Row 0 has max sum of  177


`-` 리스트 정렬

In [94]:
lst = [[1,2],[1,1],[3,1],[2,5]]
lst

[[1, 2], [1, 1], [3, 1], [2, 5]]

In [95]:
lst.sort()
lst

[[1, 1], [1, 2], [2, 5], [3, 1]]

### 2차원 리스트와 함수
2차원 리스트도 다른 객체와 마찬가지로 함수에 인수로 전달할 수도 있고, 함수가 반환값으로 반환할 수도 있다.

In [100]:
def getMatrix():
    matrix = []
    numRows = eval(input("Number of Rows: "))
    numCols = eval(input("Number of Colunmns: "))
    for row in range(numRows):
        matrix.append([])
        for col in range(numCols):
            value = eval(input("value:"))
            matrix[row].append(value)
    return matrix

In [102]:
getMatrix()

Number of Rows:  3
Number of Colunmns:  2
value: 79
value: 98
value: 46
value: 15
value: 17
value: 51


[[79, 98], [46, 15], [17, 51]]

In [103]:
def accumulate(m):
    total = 0
    for row in m:
        total += sum(row)
    return total

In [105]:
accumulate(matrix)

336

- matrix의 모든 원소의 합은 336으로 위에서 구한 (2차원 리스트 처리: 모든 원소의 합)에서 구한 값과 같다.

### 예제1 | 가장 가까운 두 점은?

여러 점에 대한 좌표가 있다. 이들 점 중에서 가장 가까운 두 점을 찾아보자.

In [1]:
def distance(x1,y1,x2,y2):
    return((x1-x2)**2 + (y1-y2)**2)**0.5

def nearestPoint(points):
    p1, p2 = 0, 1
    shortestDist = distance(points[p1][0], points[p1][1], points[p2][0], points[p2][1])
    for i in range(len(points)):
        for j in range(i+1, len(points)):
            d = distance(points[i][0], points[i][1], points[j][0], points[j][1])
            if d < shortestDist:
                shortestDist = d
                p1, p2 = i, j
    return p1, p2  

In [3]:
nPoints = eval(input("점의 수:"))
points = []
for i in range(nPoints):
    point = [0,0]
    point[0],point[1] = eval(input("좌표:"))
    points.append(point)
    
p1, p2 = nearestPoint(points)
print("(", points[p1][0], points[p1][1],")","(",points[p2][0], points[p2][1],")")

점의 수: 4
좌표: 0,0
좌표: 0,1
좌표: 0,2
좌표: 0,6


( 0 0 ) ( 0 1 )


In [5]:
nPoints = eval(input("점의 수:"))
points = []
for i in range(nPoints):
    point = [0,0]
    point[0],point[1] = eval(input("좌표:"))
    points.append(point)
    
p1, p2 = nearestPoint(points)
print("(", points[p1][0], points[p1][1],")","(",points[p2][0], points[p2][1],")")

점의 수: 4
좌표: 1,1
좌표: 2,5
좌표: 3,8
좌표: 4,14


( 2 5 ) ( 3 8 )


### 예제2 | Sudoku

![](https://upload.wikimedia.org/wikipedia/commons/3/31/Sudoku-by-L2G-20050714_solution.svg)

[게임 규칙]
- 각 열, 각 행에 1~9까지 숫자가 들어가야 한다.
- $3 \times 3$ 블록에 1~9가지 숫자가 들어가야 한다.

#### 어떤 Sudoku 해결방법이 맞는지 검사

- 두가지 검사방법

    - 1. 각 행, 열, 블록이 1~9까지 숫자를 포함하고 있는지 검사
    - 2. <font color='red'> 각 셀에 대해 그 셀의 숫자가 행, 열, 블록에서 유일한지 검사</font> <--- 이 방법 사용할 것임!

In [13]:
def isValid(grid):
    for i in range(9):
        for j in range(9):
            if grid[i][j] < 1 or grid[i][j] > 9 or not isValidAt(i,j,grid):
                return False
    return True

In [14]:
def isValidAt(i,j,grid):
    for column in range(9):
        if column != j and grid[i][column] == grid[i][j]:
            return False
    for row in range(9):
        if row != i and grid[row][j] == grid[i][j]:
            return False
    for row in range((i//3)*3, (j//3)*3 + 3):
        if row != i and column != j and grid[row][column] == grid[i][j]:
            return False
    return True

### 다차원 리스트

- 일반적으로 $n$개의 첨자 사용

```
m[i][j][k]
```

![](https://i.stack.imgur.com/soJYD.png)